File: mobile/src/main/java/net/mm2d/dmsexplorer/view/base/MainOptionsMenuDelegate.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.base;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.Menu;
import android.view.MenuItem;

import net.mm2d.dmsexplorer.R;
import net.mm2d.dmsexplorer.settings.Settings;
import net.mm2d.dmsexplorer.view.SettingsActivity;
import net.mm2d.dmsexplorer.view.dialog.UpdateDialog;
import net.mm2d.dmsexplorer.view.eventrouter.EventObserver;
import net.mm2d.dmsexplorer.view.eventrouter.EventRouter;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
class MainOptionsMenuDelegate implements OptionsMenuDelegate {
    @NonNull
    private final BaseActivity mActivity;
    @NonNull
    private final Settings mSettings;
    @NonNull
    private final EventObserver mUpdateAvailabilityObserver;

    MainOptionsMenuDelegate(@NonNull final BaseActivity activity) {
        mActivity = activity;
        mSettings = new Settings(activity);
        mUpdateAvailabilityObserver = EventRouter.getUpdateAvailabilityObserver(activity);
    }

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        mUpdateAvailabilityObserver.register(mActivity::invalidateOptionsMenu);
    }

    @Override
    public void onDestroy() {
        mUpdateAvailabilityObserver.unregister();
    }

    @Override
    public boolean onCreateOptionsMenu(@NonNull final Menu menu) {
        mActivity.getMenuInflater().inflate(R.menu.menu_main, menu);
        menu.findItem(R.id.action_update).setVisible(mSettings.isUpdateAvailable());
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull final MenuItem item) {
        final int id = item.getItemId();
        switch (id) {
            case android.R.id.home:
                mActivity.navigateUpTo();
                return true;
            case R.id.action_settings:
                SettingsActivity.start(mActivity);
                return true;
            case R.id.action_update:
                UpdateDialog.show(mActivity);
                return true;
        }
        return false;
    }

    @Override
    public boolean onPrepareOptionsMenu(@NonNull final Menu menu) {
        menu.findItem(R.id.action_update).setVisible(mSettings.isUpdateAvailable());
        return true;
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/view/base/MainOptionsMenuDelegate.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.base

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import net.mm2d.dmsexplorer.R
import net.mm2d.dmsexplorer.util.UpdateChecker
import net.mm2d.dmsexplorer.view.SettingsActivity
import net.mm2d.dmsexplorer.view.eventrouter.EventRouter

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal class MainOptionsMenuDelegate(
    private val activity: BaseActivity,
    menuResId: Int,
) : OptionsMenuDelegate {
    private val menuResId = if (menuResId != 0) menuResId else R.menu.main

    override fun onCreate(
        savedInstanceState: Bundle?,
    ) {
        EventRouter.observeUpdateAvailable(activity) {
            activity.invalidateOptionsMenu()
        }
    }

    override fun onDestroy() {
    }

    override fun onCreateOptionsMenu(
        menu: Menu,
    ): Boolean {
        activity.menuInflater.inflate(menuResId, menu)
        return true
    }

    override fun onOptionsItemSelected(
        item: MenuItem,
    ): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                activity.navigateUpTo()
                return true
            }

            R.id.action_settings -> {
                SettingsActivity.start(activity)
                return true
            }

            R.id.action_update -> {
                UpdateChecker.tryToUpdate(activity)
                return true
            }
        }
        return false
    }

    override fun onPrepareOptionsMenu(
        menu: Menu,
    ): Boolean {
        menu.findItem(R.id.action_update).isVisible = EventRouter.updateAvailable()
        return true
    }
}
